### ISSUE TITLE
Server down

### ISSUE THREAD

**Issue by Bakual** [joomla-extensions/jstats-plugin#2]
The plugin is trying to reach http://jstats.dongilbert.net/submit but that one is down.
I guess @dongilbert closed it for good.

So we either have to get that server side script onto own property if we want to get that plugin included eventually, or just remove the code.

**Comment by wilsonge** [comment]
The server code is forked to https://github.com/joomla-extensions/jstats-server

I'm currently doing some cleanup on it at https://github.com/wilsonge/jstats-plugin

**Comment by Bakual** [comment]
I had an (probably old) version of this plugin installed on my site. The site was down the last two days because this plugin timed out while waiting on a response from Dons server. Finding out what the heck was going on was a bit tricky :)
Lucky enough it was just my personal homepage which only my wife visits regulary :smile:

**Comment by wilsonge** [comment]
Hmmm that's a bug in the plugin we should fix - we can't have Joomla sites going down if the Joomla retrieval is down for whatever reason

**Comment by wilsonge** [comment]
OK I just checked my site as a result of what you just said. Found the same thing, fix incoming

**Comment by wilsonge** [comment]
Fixed - I've added a catch for a RuntimeException (thrown when the server doesn't exist). I've also set the timeout on the request to 2 seconds to avoid killing people's sites if there is an issue in the request

**Comment by wilsonge** [comment]
I'm going to close this for now - we need to work out where to put the server client - but we can discuss that in the server repo :)

**Comment by dongilbert** [comment]
No, I didn't close it for good. I hosted it there for a year with no progress on getting it included in Joomla and got tired of paying for it.

**Comment by mbabker** [comment]
> I've also set the timeout on the request to 2 seconds

I think I'd suggest a little longer, in theory it should only affect one request a couple times per day IIRC.

**Comment by wilsonge** [comment]
Yah it's currently running every 12 hours - not sure if we want to make it even less frequent than that

**Comment by zero-24** [comment]
>not sure if we want to make it even less frequent than that

how often the information is changed? Only if you update to a new Joomla Version. So why not only run on per week or month? We don't have so frequent Joomla Updates :smile:

**Comment by dongilbert** [comment]
It should run every time updates are checked.

**Comment by mbabker** [comment]
I'd rather go that route and just hardcode it into that process.  Keeping the plugin approach means either adding a plugin trigger in the update check system or it has an independent schedule.

**Comment by zero-24** [comment]
hmm so e.g. via plugin event like: onAfterUpdateCheck or onBevorUpdateCheck and there catch the extension ID 700 (Joomla CMS) :smile: 

But this has the disadvantage it runs everytime we access the Backend Control Panel. (as there run a task to check the update. I think a better one could be the update itself?

**Comment by mbabker** [comment]
The quickicon does efficient context checking (https://github.com/joomla/joomla-cms/blob/staging/plugins/quickicon/joomlaupdate/joomlaupdate.php).  Plus there is update caching.  So if it were a plugin event, it would have to trigger after the cache check validates that a new check should be run against the remote server.

**Comment by dongilbert** [comment]
Ideally it's integrated in the update check so that we can include code that tests for PHP version compatibility for extensions.

**Comment by wilsonge** [comment]
@nikosdion Would you be able to come up with a solution?

**Comment by nikosdion** [comment]
I can tell you how I have already worked around the exact same issue in my own software – I don't want it to be broken if/when my server is down.

Before anything else I made sure I'm using a separate server for the stats collection than the main site / download area. It is on a different physical server, it has its own DNS (via Amazon Route 53), the works. This was the easy part.

The hard part is how do you do the stats collection without blocking the interface. The stats collection is NOT performed in the lifetime of a regular request. It's a separate controller task, loaded in a hidden IFRAME. If the stats call fails there's no problem with the page. We could do that in Joomla! by using a system plugin which adds this hidden IFRAME only when an Administrator or Super User is logged in AND format=html (front-end or back-end). We can do so by hooking onAfterRender and injecting the IFRAME before the `</body>` tag.

However, this is not enough. We can do another stats collection web call when fetching updates through the CLI script. In order to prevent a network issue from killing the script we can put the stats collection last in the script and set a timeout of 10 seconds.

Also, collecting stats every 12 hours? That's an overkill! We are releasing once every 2 months, not once every week. Therefore we can live with last week's data just fine I guess. Once a week stats collection is more than enough.

Of course we need another two places where that handy IFRAME is injected: the final page of the installation script and right after we've finished installing a Joomla! update. Actually, we don't even have to do that. The first time someone logs into their site after installation the stats are collected. No problem. As for updates, the .php file we run to finalise the installation can simply remove the last timestamp of the stats collection script, therefore triggering stats collection when Joomla! Update tells you that the update is completed.

So what about sites which do not use the CLI script and nobody logs into them? Ugh, well, we already have the opt-in Joomla! update check plugin. I guess that it could also run the stats collection as the last thing it does but do note that this might make the site appear slower to one or two pageloads per week. Knowing the utter crap you can read in the forum I'm pretty sure some moron will tell our users to disable update checks to prevent that non-issue from occurring once every blue moon. At the same time we're talking about the exact sites which are completely indifferent to us: they never update, they are "built once, forgotten forever". Do we really want to expend energy on them?

## TL;DR

* You need a plugin which injects an IFRAME for stats collection.
* You need five extra lines in the CLI fetch updates script to also collect the stats.
* Clear the last stats collection timestamp after updating Joomla!.
* The collection frequency is once a week.
* Unpublishing the plugin disables stats collection, making the feature opt-out to conform with various local legislation (e.g. EU requires call home to be opt-out)
